Client/Ecosia/AppDelegate/CompositeAppDelegate.swift

```diff
@@ -0,0 +1,57 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+
+typealias AppDelegateType = UIResponder & UIApplicationDelegate
+typealias AppDelegateLeaf = AppDelegateType
+
+final class CompositeAppDelegate: AppDelegateType {
+    
+    private let appDelegates: [AppDelegateType]
+
+    init(appDelegates: [AppDelegateType]) {
+        self.appDelegates = appDelegates
+    }
+    
+    func application(_ application: UIApplication, willFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey : Any]? = nil) -> Bool {
+        appDelegates.forEach { _ = $0.application?(application, willFinishLaunchingWithOptions: launchOptions) }
+        return true
+    }
+
+    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+        appDelegates.forEach { _ = $0.application?(application, didFinishLaunchingWithOptions: launchOptions) }
+        return true
+    }
+    
+    func applicationDidBecomeActive(_ application: UIApplication) {
+        appDelegates.forEach { _ = $0.applicationDidBecomeActive?(application) }
+    }
+    
+    func applicationWillResignActive(_ application: UIApplication) {
+        appDelegates.forEach { _ = $0.applicationWillResignActive?(application) }
+    }
+    
+    func applicationDidEnterBackground(_ application: UIApplication) {
+        appDelegates.forEach { _ = $0.applicationDidEnterBackground?(application) }
+    }
+    
+    func application(_ application: UIApplication, performActionFor shortcutItem: UIApplicationShortcutItem, completionHandler: @escaping (Bool) -> Void) {
+        appDelegates.forEach { _ = $0.application?(application, performActionFor: shortcutItem, completionHandler: completionHandler) }
+    }
+    
+    func applicationWillEnterForeground(_ application: UIApplication) {
+        appDelegates.forEach { _ = $0.applicationWillEnterForeground?(application) }
+    }
+    
+    func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void) -> Bool {
+        let userActivityRestoringAppDelegate = appDelegates.first(where: { $0 is UserActivityRestoringAppDelegate })
+        return userActivityRestoringAppDelegate?.application?(application, continue: userActivity, restorationHandler: restorationHandler) ?? false
```

Comment:
I think I understood what you mentioned in the PR description about this and following methods as being special cases, but have a possible suggestion:

Mostly thinking out loud, from what I understood, those are basically specific Leaf types that are responsible for this specific methods and no other leaf is expected to implement them, isn't that correct?
Would it make sense to exclude this methods from the `AppDelegateType` protocol and basically implement them on our main delegate? (we could still make them a different static method on another namespace enum just to split the code if wanted). Please also let me know if it is not clear and I can try and show it as a code example 😅 

I understand some tradeoffs might be needed and don't mind leaving it like it is, just crossed my mind it might be more explicit as a different behaviour and wanted to see what you think 🙂

Reply:
> Mostly thinking out loud, from what I understood, those are basically specific Leaf types that are responsible for this specific methods and no other leaf is expected to implement them, isn't that correct?
100% correct.
The reason why I didn’t want it to be on the main delegate is due to the fact that “tomorrow” we may have another composite delegate or make major changes to it. Thus without touching the concrete AppDelegate reference `EcosiaAppDelegate` for us.